server/src/main/java/org/opensearch/cluster/decommission/DecommissionAttributeMetadata.java

```diff
@@ -0,0 +1,265 @@
+import org.opensearch.common.xcontent.ToXContent;
+import org.opensearch.common.xcontent.XContentBuilder;
+import org.opensearch.common.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.EnumSet;
+import java.util.Objects;
+
+/**
+ * Contains metadata about decommission attribute
+ *
+ * @opensearch.internal
+ */
+public class DecommissionAttributeMetadata extends AbstractNamedDiffable<Custom> implements Custom {
+
+    public static final String TYPE = "decommissionedAttribute";
+
+    private final DecommissionAttribute decommissionAttribute;
+    private DecommissionStatus status;
+    public static final String attributeType = "awareness";
+
+    /**
+     * Constructs new decommission attribute metadata with given status
+     *
+     * @param decommissionAttribute attribute details
+     * @param status                current status of the attribute decommission
+     */
+    public DecommissionAttributeMetadata(DecommissionAttribute decommissionAttribute, DecommissionStatus status) {
+        this.decommissionAttribute = decommissionAttribute;
+        this.status = status;
+    }
+
+    /**
+     * Constructs new decommission attribute metadata with status as {@link DecommissionStatus#INIT}
+     *
+     * @param decommissionAttribute attribute details
+     */
+    public DecommissionAttributeMetadata(DecommissionAttribute decommissionAttribute) {
+        this(decommissionAttribute, DecommissionStatus.INIT);
+    }
+
+    /**
+     * Returns the current decommissioned attribute
+     *
+     * @return decommissioned attributes
+     */
+    public DecommissionAttribute decommissionAttribute() {
+        return this.decommissionAttribute;
+    }
+
+    /**
+     * Returns the current status of the attribute decommission
+     *
+     * @return attribute type
+     */
+    public DecommissionStatus status() {
+        return this.status;
+    }
+
+    /**
+     * Returns instance of the metadata with updated status
+     * @param newStatus status to be updated with
+     * @return instance with valid status
+     */
+    // synchronized is strictly speaking not needed (this is called by a single thread), but just to be safe
+    public synchronized DecommissionAttributeMetadata setUpdatedStatus(DecommissionStatus newStatus) {
+        // if the current status is the expected status already, we return the same instance
+        if (newStatus.equals(status)) {
+            return this;
+        }
+        // We don't expect that INIT will be new status, as it is registered only when starting the decommission action
+        switch (newStatus) {
+            case IN_PROGRESS:
+                validateAndSetStatus(DecommissionStatus.INIT, newStatus);
+                break;
+            case SUCCESSFUL:
+                validateAndSetStatus(DecommissionStatus.IN_PROGRESS, newStatus);
+                break;
+            case FAILED:
+                // we don't need to validate here and directly update status to FAILED
```

Comment:
If any future request, passing all the checks (which was presumably idempotent) and still want to FAIL the request, then I think we should mark it FAILED. Because that request would be last running here and should be honoured

Reply:
Maybe the follow up PR should help simplify the transitions